Support unaligned array join.

amosbird opened the pull request:
For left ARRAY JOIN, expand to the greatest size. If all sizes are zero,
resize to one which is the same as the old behavior, i.e. emptyArrayToSingle

For non-left ARRAY JOIN, expand to the greatest size but keep empty if all sizes are zero.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

alexey-milovidov commented on dbms/src/Interpreters/ExpressionActions.cpp:
```diff
@@ -144,8 +144,9 @@ ExpressionAction ExpressionAction::arrayJoin(const NameSet & array_joined_column
 
-    if (array_join_is_left)
-        a.function_builder = FunctionFactory::instance().get("emptyArrayToSingle", context);
+    a.function_builder = FunctionFactory::instance().get("length", context);
```
Strange name.

alexey-milovidov replied:
Better to use separate variable.

alexey-milovidov commented on dbms/src/Interpreters/ExpressionActions.cpp:
```diff
@@ -373,26 +374,39 @@ void ExpressionAction::execute(Block & block, bool dry_run) const
 
-                any_array = &typeid_cast<const ColumnArray &>(*any_array_ptr);
+                Block tmp_block{src_col, column_of_max_length, {{}, src_col.type, {}}};
+                function_arrayResize->build({src_col, column_of_max_length})->execute(tmp_block, {0, 1}, 2, rows);
```
How it will affect performance?

alexey-milovidov reviewed (changes requested):
Performance test is required.

amosbird committed:
Support unaligned array join

For left ARRAY JOIN, expand to the greatest size. If all sizes are zero,
resize to one which is the same as the old behavior, i.e. emptyArrayToSingle

For non-left ARRAY JOIN, expand to the greatest size but keep empty if all sizes are zero.

amosbird commented:
The performance degrade badly for small arrays. It makes sense to add an option then.


```
Running: array_join
[
{
    "hostname": "201.nobida.cn",
    "main_metric": "max_rows_per_second",
    "num_cores": 20,
    "num_threads": 40,
    "ram": 134768111616,
    "runs": [
        {
            "max_rows_per_second": 67098782.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t ARRAY JOIN a, b WHERE NOT ignore(a + b)"
        },
        {
            "max_rows_per_second": 33748463.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t LEFT ARRAY JOIN a, b WHERE NOT ignore(a + b)"
        },
        {
            "max_rows_per_second": 21724193.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        },
        {
            "max_rows_per_second": 21570510.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t LEFT ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        },
        {
            "max_rows_per_second": 16701070.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2, number] b FROM system.numbers) AS t ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        },
        {
            "max_rows_per_second": 16921843.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2, number] b FROM system.numbers) AS t LEFT ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        }
    ],
    "server_version": "18.14.9",
    "test_name": "array_join",
    "time": "2018-12-12 11:11:32"
}
]
```

alexey-milovidov commented:
Ok. A setting looks reasonable.

alexey-milovidov committed:
Update Settings.h
